extension/healthcheckv2extension/internal/status/aggregation.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// Status returns the Status (enum) associated with the StatusEvent
+func (ev *statusEvent) Status() component.Status {
+	return ev.status
+}
+
+// Err returns the error associated with the StatusEvent.
+func (ev *statusEvent) Err() error {
+	return ev.err
+}
+
+// Timestamp returns the timestamp associated with the StatusEvent
+func (ev *statusEvent) Timestamp() time.Time {
+	return ev.timestamp
+}
+
+type ErrorPriority int
+
+const (
+	PriorityPermanent ErrorPriority = iota
+	PriorityRecoverable
+)
+
+type aggregationFunc func(*AggregateStatus) Event
+
+// The purpose of aggregation is to ensure that the most relevant status bubbles
+// upwards in the aggregate status. This aggregation func prioritizes lifecycle
+// events (including FatalError) over PermanentError and RecoverableError
+// events. The priority argument determines the priority of PermanentError
+// events vs RecoverableError events. Lifecycle events will have the timestamp
+// of the most recent event and error events will have the timestamp of the
+// first occurrence.
+func newAggregationFunc(priority ErrorPriority) aggregationFunc {
+	permanentPriorityFunc := func(seen map[component.Status]struct{}) component.Status {
+		if _, isPermanent := seen[component.StatusPermanentError]; isPermanent {
+			return component.StatusPermanentError
+		}
+		if _, isRecoverable := seen[component.StatusRecoverableError]; isRecoverable {
+			return component.StatusRecoverableError
+		}
+		return component.StatusNone
+	}
+
+	recoverablePriorityFunc := func(seen map[component.Status]struct{}) component.Status {
+		if _, isRecoverable := seen[component.StatusRecoverableError]; isRecoverable {
+			return component.StatusRecoverableError
+		}
+		if _, isPermanent := seen[component.StatusPermanentError]; isPermanent {
+			return component.StatusPermanentError
+		}
+		return component.StatusNone
+	}
+
+	errPriorityFunc := permanentPriorityFunc
+	if priority == PriorityRecoverable {
+		errPriorityFunc = recoverablePriorityFunc
+	}
+
+	statusFunc := func(st *AggregateStatus) component.Status {
+		seen := make(map[component.Status]struct{})
+		for _, cs := range st.ComponentStatusMap {
+			seen[cs.Status()] = struct{}{}
+		}
+
+		// All statuses are the same. Note, this will handle StatusOK and StatusStopped as these two
+		// cases require all components be in the same state.
+		if len(seen) == 1 {
+			for st := range seen {
+				return st
+			}
+		}
+
+		// Handle mixed status cases
+		if _, isFatal := seen[component.StatusFatalError]; isFatal {
+			return component.StatusFatalError
+		}
+
+		if _, isStarting := seen[component.StatusStarting]; isStarting {
+			return component.StatusStarting
+		}
+
+		if _, isStopping := seen[component.StatusStopping]; isStopping {
+			return component.StatusStopping
+		}
+
+		if _, isStopped := seen[component.StatusStopped]; isStopped {
+			return component.StatusStopping
+		}
+
+		return errPriorityFunc(seen)
+	}
+
+	return func(st *AggregateStatus) Event {
```

Comment:
I replaced the two inline funcs you mentioned with a conditional. I can probably take this one step further and make the `aggregationFunc` a method on the `Aggregator`. Part of me prefers this logic to stand on its own, since it's somewhat complicated, but I'm open to whatever makes it the most readable / understandable / maintainable etc.

Reply:
a top-level static function would also work, but the current code looks good